src/ggml.c

```diff
@@ -9297,65 +9297,114 @@ static void ggml_compute_forward_add_f32(
 
-            float * dst_ptr  = (float *) ((char *) dst->data  + i03*nb3  + i02*nb2  + i01*nb1 );
-            float * src0_ptr = (float *) ((char *) src0->data + i03*nb03 + i02*nb02 + i01*nb01);
+                float * dst_ptr  = (float *) ((char *) dst->data  + i03*nb3  + i02*nb2  + i01*nb1 );
+                float * src0_ptr = (float *) ((char *) src0->data + i03*nb03 + i02*nb02 + i01*nb01);
 
-            for (int i0 = 0; i0 < ne0; i0++) {
-                float * src1_ptr = (float *) ((char *) src1->data + i13*nb13 + i12*nb12 + i11*nb11 + i0*nb10);
+                for (int i0 = 0; i0 < ne0; i0++) {
+                    float * src1_ptr = (float *) ((char *) src1->data + i13*nb13 + i12*nb12 + i11*nb11 + i0*nb10);
 
-                dst_ptr[i0] = src0_ptr[i0] + *src1_ptr;
+                    dst_ptr[i0] = src0_ptr[i0] + *src1_ptr;
+                }
+            }
+        }
+    }
+    else {
+        if (nb00 == sizeof(float) && nb0 == sizeof(float) && nb10 == sizeof(float) && ne10 == 1) {
+            for (int ir = ir0; ir < ir1; ++ir) {
+                // src1 is broadcastable across src0 and dst in i1, i2, i3
+                const int64_t i03 = ir/(ne02*ne01);
+                const int64_t i02 = (ir - i03*ne02*ne01)/ne01;
+                const int64_t i01 = (ir - i03*ne02*ne01 - i02*ne01);
+
+                const int64_t i13 = i03 % ne13;
+                const int64_t i12 = i02 % ne12;
+                const int64_t i11 = i01 % ne11;
+
+                float * dst_ptr  = (float *) ((char *) dst->data  + i03*nb3  + i02*nb2  + i01*nb1 );
+                float * src0_ptr = (float *) ((char *) src0->data + i03*nb03 + i02*nb02 + i01*nb01);
+                float * src1_ptr = (float *) ((char *) src1->data + i13*nb13 + i12*nb12 + i11*nb11);
+
+                ggml_vec_add1_f32(ne00, dst_ptr, src0_ptr, *src1_ptr);
+            }
+        } else {
+            // all are not contiguous
+            for (int ie = ie0; ie < ie1; ++ie) {
+                // src1 is broadcastable across src0 and dst in i1, i2, i3
+                const int64_t i03 = ie/(ne02*ne01*ne00);
+                const int64_t i02 = (ie - i03*ne02*ne01*ne00)/(ne01*ne00);
+                const int64_t i01 = (ie - i03*ne02*ne01*ne00 - i02*ne01*ne00);
+                const int64_t i00 = (ie - i03*ne02*ne01*ne00 - i02*ne01*ne00 - i01*ne00);
+
+                const int64_t i13 = i03 % ne13;
+                const int64_t i12 = i02 % ne12;
+                const int64_t i11 = i01 % ne11;
+                const int64_t i10 = i00 % ne10;
+
+                float * dst_ptr  = (float *) ((char *) dst->data  + i03*nb3  + i02*nb2  + i01*nb1  + i00*nb0 );
+                float * src0_ptr = (float *) ((char *) src0->data + i03*nb03 + i02*nb02 + i01*nb01 + i00*nb00);
+                float * src1_ptr = (float *) ((char *) src1->data + i13*nb13 + i12*nb12 + i11*nb11 + i10*nb10);
+
+                *dst_ptr = *src0_ptr + *src1_ptr;
             }
```

Comment:
This branch will probably be very slow - most of the computation will probably go into computing the indices